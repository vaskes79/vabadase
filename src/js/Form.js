import axios from 'axios';

// CORS
const PROXY_URL = 'https://cors-anywhere.herokuapp.com/';
const URL = 'https://us-central1-static-bp.cloudfunctions.net/addMessage ';

// SETUP AXIOS
axios.defaults.baseURL = PROXY_URL + URL;
axios.defaults.headers.common['Accept'] = 'application/json';
axios.defaults.headers.post['Content-Type'] = 'application/json';

class Form {
  constructor(
    rootSelector = '.form',
    btnActivateSelector = '.contacts__btn',
    activeModifire = 'form--active'
  ) {
    this.form = document.querySelector(rootSelector);
    this.btn = document.querySelector(btnActivateSelector);
    this.activeModifire = activeModifire;
  }

  handleActive = () => {
    this.form.classList.toggle(this.activeModifire);
  };

  validate = () => {
    let emailRegExp = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/g;
    let nameValue = this.form.querySelector('input[name="name"]').value !== '';
    let emailValue = this.form.querySelector('input[name="email"]').value;
    let themeMessageValue =
      this.form.querySelector('input[name="theme_message"]').value !== '';
    let messageValue =
      this.form.querySelector('textarea[name="message"]').value !== '';

    emailValue = emailRegExp.test(emailValue);

    return nameValue && emailValue && themeMessageValue && messageValue;
  };

  showMsg = () => {
    let msg;
    if (window.innerWidth >= 900) {
      msg = document.querySelector('.contacts__msg');
      msg.classList.add('contacts__msg--active');
    } else {
      msg = document.querySelector('.form__msg');
      msg.classList.add('form__msg--active');
    }
    msg.innerHTML = 'ваше сообщение успешно отправлено';

    setTimeout(() => {
      if (window.innerWidth >= 900) {
        msg.classList.remove('contacts__msg--active');
      } else {
        msg.classList.remove('form__msg--active');
      }
    }, 5000);
  };

  handleSendMessage = async e => {
    e.preventDefault();
    let formData = new FormData(this.form);

    if (this.validate()) {
      const data = {
        email: formData.get('email'),
        name: formData.get('name'),
        theme_message: formData.get('theme_message'),
        message: formData.get('message'),
      };

      try {
        const rawResponse = await axios.post(null, data);
        if (rawResponse.statusText === 'OK') {
          this.handleActive();
          this.form.reset();
          this.showMsg();
        }
      } catch (err) {
        console.log('axios request error', err);
      }
    }
  };

  keyboardHandlers = e => {
    let {keyCode, which} = e;
    let {handleActive, form, activeModifire} = this;
    let formIsOpen = form.classList.contains(activeModifire);
    // esc, q
    let closeKeys = [27, 81];
    let openForm = [13, 13];
    let isClose = closeKeys.includes(keyCode) || closeKeys.includes(which);
    let isContactsBtn = e.target.classList.contains('contacts__btn');
    let isOpen = openForm.includes(keyCode) || openForm.includes(which);

    if (isClose && formIsOpen) {
      form.classList.remove(activeModifire);
    }

    if (isOpen && isContactsBtn) {
      this.form.querySelector('input[name="name"]').focus();
      handleActive();
    }
  };

  setupListeners = () => {
    this.btn.addEventListener('click', this.handleActive);
    this.form.addEventListener('submit', this.handleSendMessage);
    window.addEventListener('keydown', this.keyboardHandlers);
  };

  init = () => {
    this.setupListeners();
  };
}

export default Form;
