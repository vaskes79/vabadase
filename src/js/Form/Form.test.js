import Form from './Form';
const ROOT_CLASS = 'form';
const BTN_ACTIVE = `contacts__btn`;
const ACTIVE_MODIFIRE = `${ROOT_CLASS}--active`;
let formInst = null;
let mock = null;
let valid = null;
let elemName = null;
let elemEmail = null;
let elemTheme = null;
let elemMessage = null;
// let openHandlerMock = null;

beforeEach(() => {
  formInst = new Form();
  // openHandlerMock = jest.spyOn(formInst, 'openHandler');
  mock = jest.fn();
  formInst.init();
});

afterEach(() => {
  formInst = null;
  mock = null;
  valid = null;
  // openHandlerMock = null;
});

describe(`Create instance and render`, () => {
  it(`expect that formInst instance of Form`, () => {
    expect(formInst instanceof Form).toBeTruthy();
  });

  it(`render components into DOM`, () => {
    expect(formInst).toMatchSnapshot();
  });
});

describe(`props`, () => {
  describe(`default props`, () => {
    it(`this.formInst exists has class ${ROOT_CLASS}`, () => {
      let elem = formInst.form;
      expect(elem).not.toBeNil();
      expect(elem.classList.contains(ROOT_CLASS)).toBeTrue();
    });

    it(`this.btn exists has class ${BTN_ACTIVE}`, () => {
      let elem = formInst.btn;
      expect(elem).not.toBeNil();
      expect(elem.classList.contains(BTN_ACTIVE)).toBeTrue();
    });

    it(`this.activeModifier exists has class ${ACTIVE_MODIFIRE}`, () => {
      let selector = formInst.activeModifier;
      expect(selector).not.toBeNil();
      expect(selector).toEqual(ACTIVE_MODIFIRE);
    });
  });
});

describe('methods', () => {
  const listMethods = [
    'handleActive',
    'validate',
    'showMsg',
    'handleSendMessage',
    'keyboardHandlers',
    'setupListeners',
    'init',
  ];
  listMethods.forEach(method => {
    describe(`${method}`, () => {
      it(`${method} exists`, () => {
        expect(formInst[method]).toBeFunction();
      });
      it(`${method} toBeCalled`, () => {
        formInst[method] = mock;
        formInst[method]();
        expect(formInst[method]).toBeCalled();
      });
      if (method === 'handleActive') {
        it(`toggle state from open close`, () => {
          let elem = formInst.form;
          expect(elem.classList.contains(ACTIVE_MODIFIRE)).toBeFalse();
          formInst.handleActive();
          expect(elem.classList.contains(ACTIVE_MODIFIRE)).toBeTrue();
        });
      }
      if (method === 'validate') {
        beforeEach(() => {
          elemName = formInst.form.querySelector('input[name="name"]');
          elemEmail = formInst.form.querySelector('input[name="email"]');
          elemTheme = formInst.form.querySelector(
            'input[name="theme_message"]'
          );
          elemMessage = formInst.form.querySelector('textarea[name="message"]');
          elemName.value = '   check remove space    ';
          elemEmail.value = '   some@email.com    ';
          elemTheme.value = '  some text     ';
          elemMessage.value = '    some largest text   ';
        });

        it(`check valid data`, () => {
          valid = formInst.validate();
          expect(valid).toBeTrue();
        });

        it(`check invalid data wrong email address`, () => {
          elemEmail.value = '   some@email    ';
          valid = formInst.validate();
          expect(valid).toBeFalse();
        });
      }
    });
  });
});

describe(`events`, () => {});
