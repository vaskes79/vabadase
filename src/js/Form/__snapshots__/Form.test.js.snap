// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create instance and render render components into DOM 1`] = `
Form {
  "activeModifier": "form--active",
  "btn": <button
    class="contacts__btn"
  >
    <svg
      class="va-icon"
    >
       
      <use
        xlink:href="#va-icon-envelope"
      />
    </svg>
  </button>,
  "form": <form
    class="form"
  >
    <h3
      class="heading-3"
    >
      отправить сообщение
    </h3>
    <div
      class="form__msg"
    >
       
    </div>
    <input
      class="form__input"
      name="name"
      placeholder="имя"
      type="text"
    />
    <input
      class="form__input"
      name="email"
      placeholder="эл. почта"
      type="email"
    />
    <input
      class="form__input"
      name="theme_message"
      placeholder="тема сообщения"
      type="text"
    />
    <textarea
      class="form__textarea"
      name="message"
      placeholder="сообщение"
      rows="4"
    />
    <button
      class="form__btn"
    >
      отправить
    </button>
  </form>,
  "handleActive": [Function],
  "handleSendMessage": [Function],
  "init": [Function],
  "keyboardHandlers": [Function],
  "setupListeners": [Function],
  "showMsg": [Function],
  "validate": [Function],
}
`;
